src/drivers/power_monitor/ina226/ina226.cpp

```diff
@@ -381,47 +244,62 @@ INA226::collect()
 	/* read from the sensor */
 	perf_begin(_sample_perf);
 
-	ret = _bus_volatage = read(INA226_REG_BUSVOLTAGE);
-
-	if (_bus_volatage >= 0) {
-		ret = _power = read(INA226_REG_POWER);
-
-		if (_power >= 0) {
-			ret = _current = read(INA226_REG_CURRENT);
-
-			if (_current >= 0) {
-				ret = _shunt = read(INA226_REG_SHUNTVOLTAGE);
-
-				if (_shunt >= 0) {
-
-					power_monitor_s report{};
-					report.timestamp = hrt_absolute_time();
-					report.voltage_v = (float) _bus_volatage * INA226_VSCALE;
-					report.current_a = (float) _current * _current_lsb;
-					report.power_w   = (float) _power * _power_lsb;
-#if defined(INA226_RAW)
-					report.rconf  = read(INA226_REG_CONFIGURATION);
-					report.rsv    = read(INA226_REG_SHUNTVOLTAGE);
-					report.rbv    = read(INA226_REG_BUSVOLTAGE);
-					report.rp     = read(INA226_REG_POWER);
-					report.rc     = read(INA226_REG_CURRENT);
-					report.rcal   = read(INA226_REG_CALIBRATION);
-					report.me     = read(INA226_REG_MASKENABLE);
-					report.al     = read(INA226_REG_ALERTLIMIT);
-#endif
-
-					/* publish it */
-					_power_monitor_topic.publish(report);
-
-					ret = OK;
-					perf_end(_sample_perf);
-					return ret;
-				}
-			}
-		}
+	_bus_volatage = read(INA226_REG_BUSVOLTAGE);
+	_power = read(INA226_REG_POWER);
+	_current = read(INA226_REG_CURRENT);
+	_shunt = read(INA226_REG_SHUNTVOLTAGE);
+
+	parameter_update_s param_update;
+
+	if (_parameters_sub.copy(&param_update)) {
+		// Currently, this INA226 driver doesn't really use ModuleParams. This call to updateParams() is just to
+		// update the battery, which is registered as a child.
+		updateParams();
+	}
+
+	// Note: If the power module is connected backwards, then the values of _power, _current, and _shunt will
+	//  be negative but otherwise valid. This isn't important, because why should we support the case where
+	//  the power module is used incorrectly?
+	if (_bus_volatage >= 0 && _power >= 0 && _current >= 0 && _shunt >= 0) {
+
+		_actuators_sub.copy(&_actuator_controls);
+		_armed_sub.copy(&_armed);
+
+		/* publish it */
+		_battery.updateBatteryStatus(
+			hrt_absolute_time(),
+			(float) _bus_volatage * INA226_VSCALE,
+			(float) _current * _current_lsb,
+			true,
+			true, // TODO: Determine if this is the selected source
+			0,
+			_actuator_controls.control[actuator_controls_s::INDEX_THROTTLE],
+			_armed.armed,
+			true
+		);
+
+		ret = OK;
+
+	} else {
+		_battery.updateBatteryStatus(
+			hrt_absolute_time(),
+			0.0,
+			0.0,
+			false,
+			false, // TODO: selected source?
+			0,
+			0.0,
+			false,
+			true
+		);
+		ret = -1;
+		perf_count(_comms_errors);
```

Comment:
Is that a comms error? Shouldn't this be called something else?

Reply:
It _can be_ a comms error. When the battery is disconnected, the INA226 loses power, so you can't communicate properly with it, and end up here. However, if something else goes wrong with the INA226 or the battery, such that you can still communicate with it, but get negative values here, you still end up here. I'm not sure how to differentiate between those two cases.

Reply:
Ah ok, fine then.